### PR TITLE
feat(one-time backup): support user specified backup path

### PR DESCRIPTION
**Pull request by zhangyifan27**
This patch added a new optional field `backup_path` to `start_backup_app_request` so that users could backup an app to a specific path.

For local_service, suppose we had a configuration:
```
[[block_service.local_service]]
    type = local_service
    args = /usr/pegasus
```
If we specify `backup` as backup_path when use one-time backup, the previous backup directory is `/usr/pegasus`, now the backup directory is `/usr/pegasus/backup`. If we want to backup app to a totally user specified path, we should config args as '/'. 

For hdfs_service, the following configurations ard recommended, so that we could only constructed one `hdfsFS` object for the same namenode.
```
[[block_service.hdfs_tjwq]]
    type = hdfs_service
    args = hdfs://tjwqstaging-hdd

[[block_service.hdfs_zjy]]
    type = hdfs_service
    args = hdfs://zjyprc-hadoop
```
And if we use this configuration, users should specify a backup path every time they want to backup tables to hdfs, otherwise backup files would uploaded to the root path '/'. If both hdfs_service backup path and user specified backup path are set, we would not start a backup, and suggest users to modify the configuration.

I have tested this on onebox cluster and I will add function tests in the following patchs.

### CODE REVIEW

File: src/block_service/hdfs/hdfs_service.cpp [hunk 1]
```diff
@@ -67,11 +67,13 @@ hdfs_service::~hdfs_service()
 
 error_code hdfs_service::initialize(const std::vector<std::string> &args)
 {
-    if (args.size() != 2) {
+    if (args.size() < 1) {
         return ERR_INVALID_PARAMETERS;
     }
     _hdfs_name_node = args[0];
-    _hdfs_path = args[1];
+    _hdfs_path = args.size() >= 2 ? args[1] : "/";
```
Comment (neverchanje) [inline src/block_service/hdfs/hdfs_service.cpp:74]: Please add a comment here: what do the arguments mean separately?

**Review by Shuo-Jia — DISMISSED** [review]

File: src/meta/backup_engine.cpp [hunk 2]
```diff
@@ -71,6 +72,12 @@ error_code backup_engine::set_block_service(const std::string &provider)
     return ERR_OK;
 }
 
+void backup_engine::set_backup_path(const std::string &path)
+{
+    ddebug_f("backup path is set to {}.", path);
```
Comment (Smityz) [inline src/meta/backup_engine.cpp:77]: Why add a dot in the end?

**Comment by levy5307** [comment]
```
 If we want to backup app to a totally user specified path, we shouldn't config block_service root path or configure it as '/'. 
```
Should we add a judgement to avoid this situation?

File: src/meta/backup_engine.cpp (hunk 2 above)
Comment (zhangyifan27) [reply src/meta/backup_engine.cpp:77]: It's just a ending period, sometimes this could be omitted for brevity.

**Comment by zhangyifan27** [comment]
> ```
>  If we want to backup app to a totally user specified path, we shouldn't config block_service root path or configure it as '/'. 
> ```
> 
> Should we add a judgement to avoid this situation?



> ```
>  If we want to backup app to a totally user specified path, we shouldn't config block_service root path or configure it as '/'. 
> ```
> 
> Should we add a judgement to avoid this situation?

I didn't remove the default root path configured in `block_service` because I want to maintain compatibility with `policy backup`, which doesn't have `backup_path` field. And also this field is optional, so maybe sometimes users also want to use the configured default path.

**Comment by levy5307** [comment]
> > ```
> >  If we want to backup app to a totally user specified path, we shouldn't config block_service root path or configure it as '/'. 
> > ```
> > 
> > 
> > Should we add a judgement to avoid this situation?
> 
> I didn't remove the default root path configured in `block_service` because I want to maintain compatibility with `policy backup`, which doesn't have `backup_path` field. And also this field is optional, so maybe sometimes users also want to use the configured default path.

I understood. What I mean is that if someone wants to specify user path, we should check if the block_service root path is configured or not.

**Review by levy5307 — DISMISSED** [review]

**Review by neverchanje — APPROVED** [review]

**Review by levy5307 — APPROVED** [review]

**Pull request merged**